### PR TITLE
[Improvement-#11608][task-plugin] New time variables are added to facilitate business development

### LINKED ISSUE #11608: [Improvement][task-plugin] New time variables are added to facilitate business development
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

In the process of data development, the original time variable provided by the dolphin scheduler needs to do many secondary calculations to obtain the available business time. I want to provide more convenient time variables for business development. for example:
**First day of the month**： $[month_first_day(yyyy-MM-dd,offset)]  offset = month ,such as  $[month_first_day(yyyy-MM-dd,-1)]
**Last day of the month**： $[month_last_day(yyyy-MM-dd,offset)] offset = month ,such as $[month_last_day(yyyy-MM-dd,-1)]
**First day of the week**： $[week_first_day(yyyy-MM-dd, offset)]  offset = week, such as  $[week_first_day(yyyy-MM-dd,-1)]
 **Last day of the week**: $[week_last_day(yyyy-MM-dd, offset)]  offset = week, such as  $[week_last_day(yyyy-MM-dd,-1)] 
I will also provide more time variables with business attributes

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by fengjian1129**
## Purpose of the pull request

close #11608 

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Comment by ruanwenjun** [comment]
@fengjian1129 Please create an issue first, why we need to do this change.

**Comment by fengjian1129** [comment]
@ruanwenjun hi bro,i have created a issue,link:https://github.com/apache/dolphinscheduler/issues/11608
The purpose of adding the time variable is to provide the common business time in the production environment, which will be convenient to use. The original time variable needs multiple calculations. Moreover, in the production environment, the time variables used for business development need to be configured simply and quickly

**Comment by fengjian1129** [comment]
@ruanwenjun hi bro. can you review the code I submitted.

**Comment by fengjian1129** [comment]
@EricGao888 look this pr

**Comment by fengjian1129** [comment]
@caishunfeng @zhongjiajie @zhuangchong hello , Can I submit this PR to the community?

File: docs/docs/zh/guide/parameter/built-in.md
Comment (EricGao888) [inline docs/docs/zh/guide/parameter/built-in.md:32]: Could u plz add a corresponding English doc for this feature?

File: docs/docs/zh/guide/parameter/built-in.md
Comment (EricGao888) [inline docs/docs/zh/guide/parameter/built-in.md:34]: BTW, I think there could be a better design. Instead of using `N` and converting it into different time units, I suggest we use `y` for year, `M` or month, `d` for day. For example, replace `$[add_months(yyyyMMdd,12*N)]` with `$[add(yyyyMMdd, y)]`. But this is out of the scope of this PR. I will create a new issue for further discussion.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/TimePlaceholderUtils.java
Comment (EricGao888) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/TimePlaceholderUtils.java]:
```suggestion
    private static String transformYearWeek(Date date, String format, int weekDay) {
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/TimePlaceholderUtilsTest.java
Comment (EricGao888) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/test/java/org/apache/dolphinscheduler/plugin/task/api/parser/TimePlaceholderUtilsTest.java]: Instead of writing a long method as a test case, I suggest splitting it into multiple short cases and remove the comments. For example, we could have 

```java
@Test
    public void timePlaceHolderForThisDay () { ... }
```

here.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/TimePlaceholderUtils.java
Comment (EricGao888) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/TimePlaceholderUtils.java]:
```suggestion
    public static String calculateYearWeek(String expression, Date date) {
```

`calc` is a wired shorthand. I prefer to use `calculate` here.

**Review by EricGao888 — COMMENTED** [review]
@fengjian1129 Thanks for submitting this cool feature. I add some comments.

File: docs/docs/zh/guide/parameter/built-in.md
Comment (fengjian1129) [reply docs/docs/zh/guide/parameter/built-in.md:34]: If my PR is merged, I will consider adding this design to the next PR. However, for business development, people prefer to use independent time expressions rather than one formula to produce multiple time expressions. In other words, I want to calculate the month. I only use addmonth (); I want to calculate the year. I only use addyear () and so on

Comment (EricGao888) [reply docs/docs/zh/guide/parameter/built-in.md:33]: > BTW, I think there could be a better design. Instead of using `N` and converting it into different time units, I suggest we use `y` for year, `M` or month, `d` for day. For example, replace `$[add_months(yyyyMMdd,12*N)]` with `$[add(yyyyMMdd, y)]`. But this is out of the scope of this PR. I will create a new issue for further discussion.

+1

Comment (EricGao888) [reply docs/docs/zh/guide/parameter/built-in.md:34]: > If my PR is merged, I will consider adding this design to the next PR. However, for business development, people prefer to use independent time expressions rather than one formula to produce multiple time expressions. In other words, I want to calculate the month. I only use addmonth (); I want to calculate the year. I only use addyear () and so on

+1

**Comment by fengjian1129** [comment]
@EricGao888 Code has been updated

**Comment by EricGao888** [comment]
@ruanwenjun @caishunfeng @SbloodyS Could u plz help take a look when available? Thanks : )

**Review by davidzollo — APPROVED** [review]
LGTM
+1

**Comment by fengjian1129** [comment]
@EricGao888 @davidzollo thanks bro

**Review by EricGao888 — APPROVED** [review]
I've restarted failed jobs in CI, LGTM once CI passes.

**Comment by fengjian1129** [comment]
> I've restarted failed jobs in CI, LGTM once CI passes.

What do u want me to do ?

**Comment by EricGao888** [comment]
> > I've restarted failed jobs in CI, LGTM once CI passes.
> 
> What do u want me to do ?

Nothing actually, just wait for the CI to complete. It seems CI is not very stable recently.

**Comment by fengjian1129** [comment]
> Nothing actually, just wait for the CI to complete. It seems CI is not very stable recently.

okay, i see. Thanks

**Comment by EricGao888** [comment]
@fengjian1129 Good job! This is a handy feature, thanks again for the contribution!

**Pull request merged**
